### ISSUE TITLE
Gallery attachment without srcset attribute showing empty srcset

### ISSUE THREAD

**Issue by esolix** [ampproject/amp-wp#3087]
Wordpress Version: Version 5.2.2
AMP Version: 1.2.1 (Upgraded from 1.2.0)
Mode: Reader mode

Other Plugins Installed:
WP-Super-Cache
All-In-One-SEO
AMP
Contact Form 7
Jetpack
Classic Editor

(All plugins are up to date)

After upgrading, we are getting many empty srcset on our older posts that has no attachment attributes. Search console threw out Missing URL for attribute 'srcset' in tag 'amp-img'.

I can reproduced this on 2 sites (Both runs on different themes) - For now we have downgraded back to 1.2.0

Output on AMP version 1.2.0

```html
<amp-carousel width="600" height="414" type="slides" layout="responsive">
	<a href="/wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-2.jpg">
		<amp-img src="/wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-2.jpg" width="600" height="369" layout="responsive"></amp-img>
	</a>
	<a href="/wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-3.jpg">
		<amp-img src="/wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-3.jpg" width="600" height="350" layout="responsive"></amp-img>
	</a>
	<a href="/wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-5.jpg">
		<amp-img src="/wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-5.jpg" width="600" height="350" layout="responsive"></amp-img>
	</a>
	<a href="/wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-6.jpg">
		<amp-img src="/wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-6.jpg" width="600" height="244" layout="responsive"></amp-img>
	</a>
	<a href="/wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-9.jpg">
		<amp-img src="/wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-9.jpg" width="600" height="414" layout="responsive"></amp-img>
	</a>
</amp-carousel>
```

Output on AMP version 1.2.1 

```html
<amp-carousel width="600" height="244" type="slides" layout="responsive">
	<a href="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-2.jpg">
		<amp-img src="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-2.jpg" srcset="" width="600" height="369" layout="responsive" alt></amp-img>
	</a>
	<a href="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-3.jpg">
		<amp-img src="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-3.jpg" srcset="" width="600" height="350" layout="responsive" alt></amp-img>
	</a>
	<a href="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-5.jpg">**
		<amp-img src="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-5.jpg" srcset="" width="600" height="350" layout="responsive" alt></amp-img>
	</a>
	<a href="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-6.jpg">
		<amp-img src="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-6.jpg" srcset="" width="600" height="244" layout="responsive" alt></amp-img>
	</a>
	<a href="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-9.jpg">
		<amp-img src="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-9.jpg" srcset="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-9.jpg 600w, wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-9-555x383.jpg 555w" width="600" height="414" layout="responsive" alt></amp-img>
	</a>
</amp-carousel>
```

In 1.2.1 this line has an empty srcset

```html
<amp-img src="wp-content/uploads/2018/11/Nokia-8.1-Marketing-material-6.jpg" srcset="" width="600" height="244" layout="responsive" alt></amp-img>
```

Tried filtering out those with wp_get_attachment_image_attributes but no luck

```php
add_filter( 'wp_get_attachment_image_attributes', function( $attributes ) {
	if ( function_exists( 'is_amp_endpoint' ) && is_amp_endpoint() && empty( $attributes['srcset'] ) ) {
		unset( $attributes['srcset'] );
	}
	return $attributes;
} );
```

**Comment by westonruter** [comment]
Please share the `gallery` shortcode being used here.

Also, are the URLs actually relative like that or did you strip the host names? Is the Photon module from Jetpack active?

**Comment by esolix** [comment]
> Please share the `gallery` shortcode being used here.
> 
> Also, are the URLs actually relative like that or did you strip the host names? Is the Photon module from Jetpack active?

Sorry, I stripped out the hostname, but if needed I can provide you the URL in private if required.

We use [gallery] shortcode without any additional strings on some and some other with ids string such as [gallery ids="371643,371645,371644"]

We do not use photon, external CDN (parse replace with wp-super-cache CDN option)

**Comment by westonruter** [comment]
To confirm, the 1.2.1 output you shared includes `**`:

> `…Nokia-8.1-Marketing-material-5.jpg">**<amp-img src=…`

Is this actually in the output or was it added accidentally when copying?

**Comment by westonruter** [comment]
Also, in a gallery of 5 images, 4 of them have empty `srcset` but 1 does have it. Is there any difference between these images?

Here's the pull request which introduces the `srcset` for images in galleries: #2864. 

Note this line:

https://github.com/ampproject/amp-wp/blob/c0486f39b316092ab102424a001c3ceb21bd0bc0/includes/embeds/class-amp-gallery-embed.php#L128

Can you determine why `wp_get_attachment_image_srcset()` is returning an empty string for these 4 images? Perhaps all of the image sizes have not been generated?

The quick fix here would be to only include the `srcset` if it is actually populated, doing something like this:

```diff
diff --git a/includes/embeds/class-amp-gallery-embed.php b/includes/embeds/class-amp-gallery-embed.php
index 00f1c52f..d9f85de0 100644
--- a/includes/embeds/class-amp-gallery-embed.php
+++ b/includes/embeds/class-amp-gallery-embed.php
@@ -217,12 +217,14 @@ class AMP_Gallery_Embed_Handler extends AMP_Base_Embed_Handler {
 		foreach ( $args['images'] as $props ) {
 			$image_atts = [
 				'src'    => $props['url'],
-				'srcset' => $props['srcset'],
 				'width'  => $props['width'],
 				'height' => $props['height'],
 				'layout' => 'responsive',
 				'alt'    => $props['alt'],
 			];
+			if ( ! empty( $props['srcset'] ) ) {
+				$image_atts['srcset'] = $props['srcset'];
+			}
 
 			$this_aspect_ratio = $props['width'] / $props['height'];
 			if ( $this_aspect_ratio > $max_aspect_ratio ) {
```

**Comment by westonruter** [comment]
@esolix Please test #3089.

**Comment by esolix** [comment]
> Also, in a gallery of 5 images, 4 of them have empty `srcset` but 1 does have it. Is there any difference between these images?
> 
> Here's the pull request which introduces the `srcset` for images in galleries: #2864.
> 
> Note this line:
> 
> https://github.com/ampproject/amp-wp/blob/c0486f39b316092ab102424a001c3ceb21bd0bc0/includes/embeds/class-amp-gallery-embed.php#L128
> 
> Can you determine why `wp_get_attachment_image_srcset()` is returning an empty string for these 4 images? Perhaps all of the image sizes have not been generated?
> 
> The quick fix here would be to only include the `srcset` if it is actually populated, doing something like this:
> 
> ```diff
> diff --git a/includes/embeds/class-amp-gallery-embed.php b/includes/embeds/class-amp-gallery-embed.php
> index 00f1c52f..d9f85de0 100644
> --- a/includes/embeds/class-amp-gallery-embed.php
> +++ b/includes/embeds/class-amp-gallery-embed.php
> @@ -217,12 +217,14 @@ class AMP_Gallery_Embed_Handler extends AMP_Base_Embed_Handler {
>  		foreach ( $args['images'] as $props ) {
>  			$image_atts = [
>  				'src'    => $props['url'],
> -				'srcset' => $props['srcset'],
>  				'width'  => $props['width'],
>  				'height' => $props['height'],
>  				'layout' => 'responsive',
>  				'alt'    => $props['alt'],
>  			];
> +			if ( ! empty( $props['srcset'] ) ) {
> +				$image_atts['srcset'] = $props['srcset'];
> +			}
>  
>  			$this_aspect_ratio = $props['width'] / $props['height'];
>  			if ( $this_aspect_ratio > $max_aspect_ratio ) {
> ```

For some reason, some .jpgs doesnt get sizes within array queries from wp_get_attachment_metadata - Not sure if it is an image problem or wp uploader werent injecting those data, I will analyze the .jpg for now to see what is causing sizes data to failed.

In the mean time I will try the patch #3089

**Comment by esolix** [comment]
I can confirmed #3089 fixes this issue - Thanks!

**Comment by esolix** [comment]
> @esolix Now available: [1.2.2-RC1](https://github.com/ampproject/amp-wp/releases/tag/1.2.2-RC1).

THX!
